Feature/login-user-crud

SwanyCastle opened the pull request:
### Pull Request

> ### 제목
> - Fix: Entity 이름 변경 
> - Feat: 로그인 기능추가
> - Feat: 회원 조회, 수정, 삭제 기능 구현
> - Fix: 로그인시 토큰 Redis에서 관리 하도록 수정, 토큰이 필요한 요청을 받을시 Redis 에 해당 토큰이 존재하는지 유효성 검사 추가
> - Feat: 로그아웃 기능 구현
> - Style: 코드 스타일에 맞게 코드 정렬
> - Feat: OAuth 2.0 를 활용하여 소셜 로그인 구현 (KaKao, Naver)]

> ### 변경사항
> 
> 📌 **AS-IS**
>
> PR 리뷰 코멘트 반영 및 중복 코드 제거및 응답 메시지 수정
> 이메일 인증, 회원가입 및 Redis 데이터 생성, 조회, 삭제, 인증번호 검증, 이메일 인증여부 확인 테스트 코드 작성
>
> 🔖 **TO-BE**
>
> 🔧  Entity 이름 변경 및 로그인 기능추가
> - Entity 이름을 User 에서 Member 로 변경
> - 기본 로그인 기능 추가
>
> 🔧  회원 조회, 수정, 삭제 기능 구현 및 테스트코드 작성
>   - 요청으로 들어온 Member Id 에대한 유저가 존재하는지 확인 후 조회, 수정, 삭제 처리하도록 구현
>   - 로그인 및 회원 조회, 수정, 삭제 기능에 대한 테스트 코드 작성
>
> 🔧  로그인시 토큰 Redis에서 관리 하도록 수정, 토큰이 필요한 요청을 받을시 Redis 에 해당 토큰이 존재하는지 토큰 유효성 검사 추가
> - 로그인시 토큰을 Redis 에 서 관리 하도록 수정
>    - 로그인 할 때 토큰생성을 기존의 Controller Layer 에서 Service Layer 에서 생성하도록 변경
>    - 토큰 생성 후 생성된 토큰을 유효시간 설정하여 Redis 에 memberId 를 키값으로해 저장
>
> - 토큰이 필요한 요청을 받을때 Redis 에 해당 토큰이 존재하는지 유효성 검사 추가
>    - 토큰이 존재한다면 정상 진행 존재하지 않다면 No Permission 에러 응답
>
> 🔧  로그아웃 기능 구현
> - 로그아웃시 토큰이 있어야지만 접근가능하도록 설정
> - 로그아웃시 Path Variable 로 넘어온 memberId 와 토큰안에 있는 memberId 의 값을 비교해 사용자 본인이 요청한것인지 확인
> - 확인 완료 후 Redis 에 있는 memberId 를 키 값으로 하는 토큰을 삭제
>
>  🔧  코드 스타일에 맞게 코드 정렬
>
> 🔧  OAuth 2.0 를 활용하여 소셜 로그인 구현
> - application.yml 에 OAuth 2.0 사용을 위한 Kakao, Naver 의 설정 및 각 provider 설정
> - OAuth 2.0 사용을 위해 Security Filter Chain 설정에 oauth2Login 설정
> 	- authorizationEndpoint() 를 사용해 Kakao 나 Naver 의 인증서버로 Redirection
> 	- 로그인을 완료하면 설정한 redirectionEndpoint 로 엑세스토큰을 포함한 Redirection
> 	- Spring Security OAuth 2.0 에서  이 요청을 받아 userInfoEndpoint 에서 설정한 oAuth2UserService 와 Kakao 나 Naverr 로 부터  받은 엑세스토큰을 사용하여 요청해서 사용자 정보 조회
>   - OAuth2MemberServiceImplement 의 인스턴스인  oAuth2UserService 의 loadUser() 메서드가 호출
> 	- Ecommerce Application 내의 DB에 소셜 로그인한 사용자의 정보를 저장
> 	- Ecommerce Application 의 토큰을 생성해 발급해주기 위해 memberId 값을 반환
>         - oAuth2SuccessHandler 에서 loadUser() 에서 반환한  memberId 값으로 토큰을 생성해 Redis 에서 관리하고 생성된 토큰을 응답

> ### 테스트
>
> - [x] 테스트 코드 ( 기본 로그인, 회원 조회, 수정, 삭제 )
> - [x] API 테스트

SwanyCastle committed:
[Fix]: Entity 이름 변경 및 로그인 기능추가

- Entity 이름을 User 에서 Member 로 변경
- 로그인 기능 추가 및 API TEST 확인

SwanyCastle committed:
[Feat]: 로그인 및 회원 조회, 수정, 삭제 기능 구현
[Test]: 로그인 및 회원 조회, 수정, 삭제 기능에 대한 테스트 코드 작성

로그인 기능 구현
- 로그인시 비밀번호 일치 확인후 토큰 발급하여 응답하도록 구현
- 로그인 기능에 대한 테스트 코드 작성

회원 조회, 수정, 삭제 기능 구현
- 요청으로 들어온 Member Id 에대한 유저가 존재하는지 확인 후 조회, 수정, 삭제 처리하도록 구현
- 회원 조회, 수정, 삭제 기능에 대한 테스트 코드 작성

SwanyCastle committed:
[Fix]: 로그인시 토큰 Redis에서 관리 하도록 수정, 토큰이 필요한 요청을 받을시 Redis 에 해당 토큰이 존재하는지 유효성 검사 추가
[Feat]: 로그아웃 기능 구현

로그인시 토큰을 Redis 에 서 관리 하도록 수정
- 로그인 할 때 토큰생성을 기존의 Controller Layer 에서 Service Layer 에서 생성하도록 변경
- 토큰 생성 후 생성된 토큰을 유효시간 설정하여 Redis 에 memberId 를 키값으로해 저장

토큰이 필요한 요청을 받을때 Redis 에 해당 토큰이 존재하는지 유효성 검사 추가
- 토큰이 존재한다면 정상 진행 존재하지 않다면 No Permission 에러 응답

로그아웃 기능 구현
- 로그아웃시 토큰이 있어야지만 접근가능하도록 설정
- 로그아웃시 Path Variable 로 넘어온 memberId 와 토큰안에 있는 memberId 의 값을 비교해 사용자 본인이 요청한것인지 확인
- 확인 완료 후 Redis 에 있는 memberId 를 키 값으로 하는 토큰을 삭제

SwanyCastle committed:
[Style]: 코드 스타일에 맞게 코드 정렬

SwanyCastle committed:
[Feat] : OAuth 2.0 를 활용하여 소셜 로그인 구현 (KaKao, Naver)

OAuth 2.0 를 활용하여 소셜 로그인 구현
- application.yml 에 OAuth 2.0 사용을 위한 Kakao, Naver 의 설정 및 각 provider 설정
- OAuth 2.0 사용을 위해 Security Filter Chain 설정에 oauth2Login 설정
	- authorizationEndpoint() 를 사용해 Kakao 나 Naver 의 인증서버로 Redirection
	- 로그인을 완료하면 설정한 redirectionEndpoint 로 엑세스토큰을 포함한 Redirection
	- Spring Security OAuth 2.0 에서  이 요청을 받아 userInfoEndpoint 에서 설정한 oAuth2UserService 와 Kakao 나 Naverr 로 부터 받은 엑세스토큰을 사용하여 요청해서 사용자 정보 조회
        - OAuth2MemberServiceImplement 의 인스턴스인  oAuth2UserService 의 loadUser() 메서드가 호출
	- Ecommerce Application 내의 DB에 소셜 로그인한 사용자의 정보를 저장
	- Ecommerce Application 의 토큰을 생성해 발급해주기 위해 memberId 값을 반환
        - oAuth2SuccessHandler 에서 loadUser() 에서 반환한  memberId 값으로 토큰을 생성해 Redis 에서 관리하고 생성된 토큰을 응답

0xc0debabe commented:
OAuth2.0을 활용하셔서 구현하셨네요
저도 써보고 싶은 기술이였는데 잘봤습니다👍🏻

hjkim22 commented:
로그아웃기능구현 + 레디스 활용 코드잘봤습니다
OAuth 활용 소셜로그인도 자세히보고 참고해봐야겠어요 수고하셨습니다!

spring-kotlin-lover commented on src/main/java/com/ecommerce/service/auth/OAuth2MemberServiceImplement.java:
```diff
@@ -0,0 +1,87 @@
+    OAuth2User oAuth2User = super.loadUser(request);
+    String oauthClientName = request.getClientRegistration().getClientName();
+
+    ObjectMapper objectMapper = new ObjectMapper();
```
objectMapper 는 반드시 의존성 주입해서 사용하거나, 특정 client 에 종속되는 objectMapper 를 사용할 것이라면 클래스 멤버변수로 정의해서 사용해주세요~!

spring-kotlin-lover commented on src/main/java/com/ecommerce/service/auth/OAuth2MemberServiceImplement.java:
```diff
@@ -0,0 +1,87 @@
+      }
+      default -> throw new OAuth2AuthenticationException(
+          ResponseCode.UNSUPPORTED_OAUTH_PROVIDER.getDescription());
+    }
```
member 를 처음에 null 로 정의하지 말고, non-null 하게 만들어볼 수 있을 것 같은데요. 이 로직들을 private 함수로 정의해보는 것은 어떨까요?

SwanyCastle committed:
[Fix]: PR 리뷰 코멘트 사항 수정

- objectMapper 를 new 로 객체 생성하지 않고 의존성 주입하여 사용
- member 를 null 로 초기화 하지않고 관련된 해당 부분을 createSnsMember 메서드로 만들어서
  non-null 하게 수정
